kube/egressservices/egressservices.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package egressservices contains shared types for exposing tailnet services to
+// cluster workloads.
+// These are split into a separate package for consumption of
+// non-Kubernetes shared libraries and binaries. Be mindful of not increasing
+// dependency size for those consumers when adding anything new here.
+package egressservices
+
+import (
+	"net/netip"
+)
+
+// KeyEgressServices is name of the proxy state Secret field that contains the
+// currently applied egress proxy config.
+const KeyEgressServices = "egress-services"
+
+// Configs contains the desired configuration for egress services keyed by
+// service name.
+type Configs map[string]Config
+
+// Config is an egress service configuration.
+type Config struct {
+	// TailnetTarget is the target to which cluster traffic for this service
+	// should be proxied.
+	TailnetTarget TailnetTarget `json:"tailnetTarget"`
+	// Ports contains mappings for ports that can be accessed on the tailnet
+	// target keyed by a predictable name for easier lookup.
+	// {"tcp:80:4003":{"protocol":"tcp","src":80,"dst":4003}}
+	Ports map[PortMapName]PortMap `json:"ports"`
```

Comment:
> Both PortMapName and PortMap look like they have the same information in them

They do, but I don't want to loop over and compare array items to find a port- especially in the follow up PRs where I need to compare a port as defined here with a port as it is on a Service spec with additional fields potentially added by kube- so lookup by name that includes things we care about makes it easier

Reply:
Sorry I probably didn't explain this well - you can use the `PortMap` type as the map key instead of having to create a string-type purely for the purpose of being the map key, because the `PortMap` type is a comparable struct. I pushed 005b2e2f22db4a68df9e0bb78bf74e0de0e6e93d to demonstrate how this _slightly_ simplifies the code. Very much a nit, I mostly commented because I wondered if I was missing something.

Reply:
And I don't entirely know where this is going in the next PR, so I may be missing the reason why that struct will no longer work as a key in the near future.

Reply:
@tomhjp this would be a lot nicer, but we also need to encode this as JSON and afaiu JSON does not support objects as map key types. I think it's probably possible via a custom encoder or something, but in that case I'm not sure if it's simpler any more

Reply:
Seems like I could convert the map if it had struct keys to json [via a custom `MarshalText` method for the struct](https://pkg.go.dev/encoding/json#Marshal:~:text=Map%20values%20encode%20as%20JSON%20objects.%20The%20map%27s%20key%20type%20must%20either%20be%20a%20string%2C%20an%20integer%20type%2C%20or%20implement%20encoding.TextMarshaler.). That would still be fine
But then I would also need to parse those strings back into structs when converting json back to go and that seems a bit messy

Reply:
I'll try it out though

Reply:
Yeah, with the custom marshalling this still seems like an improvement, thanks

https://github.com/tailscale/tailscale/pull/13531/commits/70fc37363d7f02eba72dc771fda74e4d31cd0804#diff-62d368cb3988ff2fc5340b71425899638d96b4df7d0551de88f3f6db4e174a17R55-R84

Reply:
Ah I missed the JSON serialisation requirement. It doesn't feel right to me that the requirements of the code leak into the requirements of the serialisation and vice-versa.

I think the serialised form is bested suited to being a simple list of objects, as that most closely represents how that config gets turned into real nftable/iptable rules, and is very extensible for any future updates we make (including adding a version field in those objects). We can separately handle converting that list of objects into a set as we parse the config file - no need to tie it up in the JSON marshalling interface.